### DISCUSSION TITLE
Should we change default python lsp from `pyright` to `pylsp`?

### DISCUSSION THREAD

**Discussion by ayamir** [ayamir/nvimdots#708] [Polls]
`pyright` has always been the default lsp server for python in `nvimdots`, but it seems have little support for machine learning libraries like `python-opencv` and others. For example, `pyright` will complain errors for below code, but `pylsp` and `jedi` can resolve them normally. `pylsp` seems have more strict check compared to `jedi`.
`pyright`:
<img width="1050" alt="image" src="https://user-images.githubusercontent.com/61657399/236407218-acc2519c-d840-43a8-8cd0-89f422356735.png">
`pylsp`:
<img width="886" alt="image" src="https://user-images.githubusercontent.com/61657399/236410719-75d5ff8f-6e8c-4328-a8b0-5ba6ed39b113.png">
`jedi`:
<img width="883" alt="image" src="https://user-images.githubusercontent.com/61657399/236430236-9d43be40-b1fb-4224-a7af-25cb1d7d2b49.png">

**Comment by charliie-dev** [comment]
just some reference materials[^1][^2][^3]

[^1]: https://www.reddit.com/r/neovim/comments/rjrytp/which_python_lsp_is_better/
[^2]: https://discuss.python.org/t/lsp-for-python/22194/8
[^3]: https://stackoverflow.com/questions/55897160/what-is-the-difference-between-jedi-and-python-language-server-in-vs-code-ide

**Comment by charliie-dev** [comment]
maybe add `jedi-language-server` as an option too?

**Reply by ayamir** [reply]
It's ok.

**Comment by charliie-dev** [comment]
side note: I recently start using https://github.com/charliermarsh/ruff to replace black, it seems a bit faster(?
Or I'm just be charmed by rust lol
Maybe we can try switch that too lol

**Reply by ayamir** [reply]
Does it have code completion or just code lint? 

> Ruff can be used to replace [Flake8](https://pypi.org/project/flake8/) (plus dozens of plugins), [isort](https://pypi.org/project/isort/), [pydocstyle](https://pypi.org/project/pydocstyle/), [yesqa](https://github.com/asottile/yesqa), [eradicate](https://pypi.org/project/eradicate/), [pyupgrade](https://pypi.org/project/pyupgrade/), and [autoflake](https://pypi.org/project/autoflake/), all while executing tens or hundreds of times faster than any individual tool.

Seems like it just a lint tool to replace flake8.

**Reply by charliie-dev** [reply]
ys it's mostly a linter and formatter.

> By default, Ruff enables Flake8's `E` and `F` rules. Ruff supports all rules from the `F` category, and a [subset](https://beta.ruff.rs/docs/rules/#error-e) of the `E` category, omitting those stylistic rules made obsolete by the use of an autoformatter, like [Black](https://github.com/psf/black).

**Reply by charliie-dev** [reply]
I only use `ruff` for linting and formatting in python nowadays. I uninstall `black`, `flake8` etc, and I'm pretty happy with it. However, it's only one man's opinion. So others might need to test it first.

**Comment by ayamir** [comment]
If no one vote for this, I will replace `pyright` with `jedi_language_server`.

**Comment by fecet** [comment]
I haven't update my neovim config for a while and it's seems that we have experience pyright->jedi->pylsp lol. 

I have only used pyright before, do we have code action on other lsp?

**Reply by charliie-dev** [reply]
Ys, `pylsp` has code actions. https://github.com/python-lsp/python-lsp-server#lsp-server-features

**Reply by charliie-dev** [reply]
I also accidentally found this interesting project that extends `pylsp`'s ability.
https://github.com/python-rope/pylsp-rope
`mason` doesn't seem to support it tho.

**Reply by fecet** [reply]
Awesome, code action is extremly helpful for me and good to know it's hopeful to have refactor power in python one day.

**Comment by fecet** [comment]
A couple of things:

1. Pylsp's third-party plugins seems can be installed by ":PylspInstall", including pylsp-rope, am I missing something?
2. Pylsp will give warnings for unused variables, instead of dim them, https://github.com/zbirenbaum/neodim/issues/21 may be relavent.
3. It will analyse whole file every I saved, I set 
   ```lua
   settings = {
		pylsp = {
			plugins = {
				jedi_completion = {
					include_params = false,
				},
				pycodestyle = { enabled = false },
				flake8 = { enabled = false },
                autopep8 = { enabled = false },
			},
		},
	}
   ```
   but doesn't work

**Reply by ayamir** [reply]
Save the file to `lua/modules/configs/completion/servers/pylsp.lua`.
```lua
return {
	settings = {
		pylsp = {
			plugins = {
				jedi_completion = {
					include_params = false,
				},
				pycodestyle = { enabled = false },
				flake8 = { enabled = false },
				autopep8 = { enabled = false },
			},
		},
	},
}
```
It works for me.

**Comment by Lraxer** [comment]
Pylsp works perfectly when I'm using virtual environment created by conda or other tools. However, when I use `/usr/bin/python` directly, the autocompletion only works for some "core modules" like `sys` and `math`, but not for the 3rd party modules like  `numpy` and `selenium`. 

Because when we are not in a virtual environment, pylsp is in its own venv `.local/share/nvim/mason/packages/python-lsp-server/venv/`. We can edit `pyvenv.cfg` and set `include-system-site-packages = true` (this param is set to `false` by default) to solve this, which won't effect other virtual environments.

Maybe we can add this to wiki? :)

**Reply by charliie-dev** [reply]
That's a great idea! Nice!

**Reply by charliie-dev** [reply]
Added.
https://github.com/ayamir/nvimdots/wiki/Knowledge-Garden#-lsp-related-issues

**Reply by Konfekt** [reply]
I didn't find much documentation on `include-system-site-packages`. 
Would these system packages also be included in a virtual environment?

**Reply by Lraxer** [reply]
> I didn't find much documentation on `include-system-site-packages`. Would these system packages also be included in a virtual environment?

Hello, if you set `include-system-site-packages=true`, it means that for those packages installed at the system level, they are also visible to this virtual environment (they are not installed again in this venv.). If you install pylsp in mason.nvim, when you have activated a venv and then start neovim, pylsp can detect packages installed in that venv, it won't use its own venv. But if you are not in any venv, pylsp will use its own venv, and system-level packages are invisible to it, so you may have to change the option in `pyvenv.cfg`, as I said above.
